modules/gdscript/gdscript_analyzer.cpp

```diff
@@ -1917,6 +1927,14 @@ void GDScriptAnalyzer::resolve_assignable(GDScriptParser::AssignableNode *p_assi
 		}
 	}
 
+#ifdef DEBUG_ENABLED
+	if (GLOBAL_GET("debug/gdscript/warnings/enforce_static_typing") && type.type_source != GDScriptParser::DataType::ANNOTATED_INFERRED) {
```

Comment:
Instead of `DataType::type_source` you should check `AssignableNode::infer_datatype`. Note that neither `datatype_specifier != nullptr` (`has_specified_type` above in `resolve_assignable()`) nor `infer_datatype` is required for constants.

https://github.com/godotengine/godot/blob/0a7f75ec7b465604b6496c8f5f1d638aed250d6d/modules/gdscript/gdscript_parser.h#L388-L392